File: deploy.rs
#!/usr/bin/env cargo +nightly -Zscript

---
[package]
edition = "2021"
version = "0.1.0"
[dependencies]
clap = { version = "4", features = ["derive"] }
cmd_lib = { version = "1.9.5" }
semver = "1.0.26"
---

use clap::Parser;
use cmd_lib::*;
use semver::Version;
use std::io;
use std::io::Write;
use std::io::Read;
use std::fs;

#[derive(Parser, Debug)]
#[clap(version)]
struct Args {
    #[clap(short, long, help = "Path to config")]
    config: Option<std::path::PathBuf>,
}

#[cmd_lib::main]
fn main() -> CmdResult {
    let _args = Args::parse();

    run_cmd! (warn "You're deploying reflective to production")?;

    let status = run_fun! (git status --porcelain)?;

    if status.len() > 0 {
        run_cmd! (error "There are uncommitted changes")?;
        //return Ok(());
    }

    let version = run_fun! (
        cd service;
        cargo metadata --format-version=1 --no-deps | jq ".packages[0].version"
    )?;

    let version = version.trim().replace('"', "");

    let version = Version::parse(&version).unwrap();

    let new_version = {
        let mut version = version.clone();
        version.patch = version.patch + 1;
        version
    };

    println!("Current version: {}", version);

    print!("Enter new version (default {}): ", new_version);
    io::stdout().flush().expect("Failed to flush stdout");

    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");

    let new_version = if input.trim().is_empty() {
        new_version
    } else {
        match Version::parse(input.trim()) {
            Ok(v) => v,
            Err(_) => {
                run_cmd! (error "Invalid version" $input)?;
                return Ok(());
            }
        }
    };

    println!("Releasing version {}", new_version);

    run_cmd! (
        cd service;
        cargo set-version "$new_version";
        cd ..;
        docker buildx build --platform=linux/amd64 -t "ghcr.io/beingflo/reflective:$new_version" .;
        docker push "ghcr.io/beingflo/reflective:$new_version";
    )?;

    let contents = fs::read_to_string("./docker-compose.prod.yml").expect("Couldn't read docker compose file");
    let contents = contents.replace(&format!("ghcr.io/beingflo/reflective:{}", version), &format!("ghcr.io/beingflo/reflective:{}", new_version));
    fs::write("./docker-compose.prod.yml", contents).expect("Couldn't write docker compose file");

    run_cmd! (
        docker --context omni compose --file docker-compose.prod.yml pull;
        docker --context omni compose --file docker-compose.prod.yml up -d;
        git commit -am "Release $new_version";
        git tag "$new_version";
        git push origin --tags;
    )?;


    Ok(())
}

File: deploy.js
#!/usr/bin/env bun
import { $ } from "bun";
import { unlinkSync } from "fs";
import { parseArgs } from "util";

// Exit on error and ensure decrypted .env file is cleaned up
$.throws(true);
process.on("exit", () => {
  try {
    unlinkSync(".env.prod");
  } catch {}
});

// Parse CLI arguments
const { values } = parseArgs({
  args: Bun.argv.slice(2),
  options: {
    build: { type: "boolean", short: "b", default: false },
    deploy: { type: "boolean", short: "d", default: false },
  },
  strict: true,
});

// If no flags specified, do both
const shouldBuild = values.build || (!values.build && !values.deploy);
const shouldDeploy = values.deploy || (!values.build && !values.deploy);

console.log("Deploying reflective to production!");

if (shouldBuild) {
  // Check for uncommitted changes
  const status = await $`git status --porcelain`.text();
  if (status.trim()) {
    console.error("There are uncommitted changes");
    //process.exit(1);
  }

  // Get current version from Cargo.toml
  const metadata = await $`cargo metadata --format-version=1 --no-deps`
    .cwd("service")
    .json();
  const version = metadata.packages[0].version;
  console.log(`Latest version: ${version}`);

  // Calculate next version
  const versionParts = version.split(".");
  versionParts[versionParts.length - 1] = String(
    Number(versionParts[versionParts.length - 1]) + 1,
  );
  const nextVersion = versionParts.join(".");

  // Prompt for version
  const newVersion =
    prompt(`Enter version to be deployed [${nextVersion}]:`)?.trim() ||
    nextVersion;

  // Set new version in Cargo.toml
  await $`cargo set-version ${newVersion}`.cwd("service");

  // Build Docker image
  const imageName = `ghcr.io/beingflo/reflective:${newVersion}`;
  await $`docker buildx build -t ${imageName} .`;
  await $`docker push ${imageName}`;

  // Update compose file
  const composeFile = await Bun.file("./compose.prod.yml").text();
  const oldImageLine = `image: "ghcr.io/beingflo/reflective:${version}"`;
  const newImageLine = `image: "ghcr.io/beingflo/reflective:${newVersion}"`;
  const updatedCompose = composeFile.replace(oldImageLine, newImageLine);
  await Bun.write("./compose.prod.yml", updatedCompose);

  // Git operations
  await $`git commit -am "Release ${newVersion}"`;
  await $`git tag ${newVersion}`;
  await $`git push`;
  await $`git push origin --tags`;
}

if (shouldDeploy) {
  // Get age key from 1Password
  const ageKey =
    await $`op item get "SOPS age key - reflective" --reveal --fields "private key"`.text();

  // Decrypt .env file
  await $`sops -d --input-type dotenv --output-type dotenv .env.prod.enc > .env.prod`.env(
    {
      ...process.env,
      SOPS_AGE_KEY: ageKey.trim(),
    },
  );

  // Pull and deploy
  await $`docker --context arm compose --file compose.prod.yml pull`;
  await $`docker --context arm compose --file compose.prod.yml up --build -d`;
}
